wremnants/theory_tools.py

```diff
@@ -180,17 +190,97 @@ def define_prefsr_vars(df):
 
     return df
 
-def define_scale_tensor(df):
-    if "scaleWeights_tensor" in df.GetColumnNames():
-        logger.debug("scaleWeight_tensor already defined, do nothing here.")
+def define_postfsr_vars(df, mode=None):
+    if "postfsrLeptons" in df.GetColumnNames():
+        logger.debug("PostFSR leptons are already defined, do nothing here.")
         return df
-    # convert vector of scale weights to 3x3 tensor and clip weights to |weight|<10.
-    df = df.Define("scaleWeights_tensor", f"wrem::makeScaleTensor(LHEScaleWeight, theory_weight_truncate);")
-    df = df.Define("scaleWeights_tensor_wnom", "auto res = scaleWeights_tensor; res = nominal_weight*res; return res;")
+
+    # status flags in NanoAOD: https://cms-nanoaod-integration.web.cern.ch/autoDoc/NanoAODv9/2016ULpostVFP/doc_TTToSemiLeptonic_TuneCP5_13TeV-powheg-pythia8_RunIISummer20UL16NanoAODv9-106X_mcRun2_asymptotic_v17-v1.html
+    # post fsr definition
+    # 1) is stable
+    # 2) isPrompt or isDirectPromptTauDecayProduct
+    # 3) is fromHardProcess
+    # 4) is lepton
+    df = df.Define("postfsrLeptons", """
+        GenPart_status == 1
+        && (GenPart_statusFlags & 1 || GenPart_statusFlags & (1 << 5))
+        && GenPart_statusFlags & (1 << 8)
+        && abs(GenPart_pdgId) >= 11 && abs(GenPart_pdgId) <= 16 
+        """)
+    df = df.Define("postfsrElectrons", "postfsrLeptons && abs(GenPart_pdgId) == 11")
+    df = df.Define("postfsrMuons", "postfsrLeptons && abs(GenPart_pdgId) == 13")
+
+    if mode is not None:
+        # defition of more complex postfsr object 
+        # use fiducial gen met, see: https://twiki.cern.ch/twiki/bin/viewauth/CMS/ParticleLevelProducer
+        if mode in ["wlike", "dilepton"]:
+            # find the leading charged lepton and antilepton idx
+            df = df.Define("postfsrLep", "postfsrLeptons && (GenPart_pdgId==11 || GenPart_pdgId==13)")
+            df = df.Define("postfsrAntiLep", "postfsrLeptons && (GenPart_pdgId==-11 || GenPart_pdgId==-13)")
+            
+            df = df.Define("postfsrLep_idx",     "ROOT::VecOps::ArgMax(GenPart_pt[postfsrLep])")
+            df = df.Define("postfsrAntiLep_idx", "ROOT::VecOps::ArgMax(GenPart_pt[postfsrAntiLep])")
+
+            df = df.Define("postfsrLep_pt",     "event % 2 == 0 ? static_cast<double>(GenPart_pt[postfsrLep][postfsrLep_idx]) : static_cast<double>(GenPart_pt[postfsrAntiLep][postfsrAntiLep_idx])")
```

Comment:
event % 2 appears several time, shouldn't we define a variable for it to avoid repeating the operation ? Like df = df.Define("isOddEvent", "event % 2")

Reply:
I don't see much simplification in this, would leave it as it is